src/Stack/Package.hs

```diff
@@ -1362,6 +1364,57 @@ findOrGenerateCabalFile pkgDir = do
             _:_ -> Left $ PackageMultipleCabalFilesFound pkgDir files
       where hasExtension fp x = FilePath.takeExtension fp == "." ++ x
 
+-- | Searches user path for hpack executables newer than bundled one.
+findUserExecutableHpack :: HasConfig env => Config -> RIO env HpackExecutable
```

Comment:
Main issue with this is that it will re-query hpack version every single time.  I think it makes sense to cache it.  The pattern of having an `IORef (Maybe ...)` in scope is used elsewhere, and I think would be good to use here.